### ISSUE TITLE
Maxout Layer

### ISSUE THREAD

**Issue by vzhong** [pytorch/pytorch#805]
Are there plans for a maxout layer? For example:

```python
class Maxout(nn.Module):

    def __init__(self, d_in, d_out, pool_size):
        super().__init__()
        self.d_in, self.d_out, self.pool_size = d_in, d_out, pool_size
        self.lin = Linear(d_in, d_out * pool_size)

    def forward(self, inputs):
        shape = list(inputs.size())
        shape[-1] = self.d_out
        shape.append(self.pool_size)
        last_dim = len(shape) - 1
        out = self.lin(inputs)
        m, i = out.view(*shape).max(last_dim)
        return m.squeeze(last_dim)
```

**Comment by soumith** [comment]
as your example itself shows, it's simple to write by the user itself.
We're not sure it's common enough to write a reference layer in the pytorch core.
For now, we're probably not going to add it, unless a lot of folks think otherwise.

**Comment by erogol** [comment]
For ones who needs Maxout, I changed the above code to make it work. 

```
class Maxout(nn.Module):

    def __init__(self, d_in, d_out, pool_size):
        super().__init__()
        self.d_in, self.d_out, self.pool_size = d_in, d_out, pool_size
        self.lin = nn.Linear(d_in, d_out * pool_size)


    def forward(self, inputs):
        shape = list(inputs.size())
        shape[-1] = self.d_out
        shape.append(self.pool_size)
        max_dim = len(shape) - 1
        out = self.lin(inputs)
        m, i = out.view(*shape).max(max_dim)
        return m
```

**Comment by kkorovesis** [comment]
Can you please explain how this works please ? I have a 2D tensor off size [128,600] and I want to get it through a Maxout Layer.

**Comment by lucasb-eyer** [comment]
And here's a version that is a pure non-linearity applied on the last dimension, this way it can be used with more than just Linears: convs of any dimension, time-series, etc. But it's almost trivial, honestly, so I agree with not including it in the lib to avoid bloat.

```python
class Maxout(nn.Module):
    def __init__(self, pool_size):
        super().__init__()
        self._pool_size = pool_size

    def forward(self, x):
        assert x.shape[-1] % self._pool_size == 0, \
            'Wrong input last dim size ({}) for Maxout({})'.format(x.shape[-1], self._pool_size)
        m, i = x.view(*x.shape[:-1], x.shape[-1] // self._pool_size, self._pool_size).max(-1)
        return m
```

Example use:

```python
torch.arange(3*6).view(3,6)
#tensor([[  0,   1,   2,   3,   4,   5],
#        [  6,   7,   8,   9,  10,  11],
#        [ 12,  13,  14,  15,  16,  17]])

Maxout(3)(torch.arange(3*6).view(3,6))
#tensor([[  2,   5],
#        [  8,  11],
#        [ 14,  17]])

**Comment by davidtvs** [comment]
Actually, Maxout is applied to the channel dimension which in PyTorch is dimension 1 and not the last:

```py
class Maxout(nn.Module):
    def __init__(self, pool_size):
        super().__init__()
        self._pool_size = pool_size

    def forward(self, x):
        assert x.shape[1] % self._pool_size == 0, \
            'Wrong input last dim size ({}) for Maxout({})'.format(x.shape[1], self._pool_size)
        m, i = x.view(*x.shape[:1], x.shape[1] // self._pool_size, self._pool_size, *x.shape[2:]).max(2)
        return m
```

**Comment by shamoons** [comment]
If ReLU6 made it into the core, I should think that max out would as well.

**Comment by paniabhisek** [comment]
[Here](https://github.com/paniabhisek/maxout/blob/master/maxout.py), you can find both variant(`Linear` and `Conv`). The `Linear` version is more efficient than presented above.

**Comment by cowwoc** [comment]
Honestly, this is a mess. TensorFlow provides a very simple API: https://www.tensorflow.org/addons/api_docs/python/tfa/layers/Maxout

Some of the implementations I see for PyTorch build Linear layers under the hood, others don't. Some of them invoke `Tensor.view()` with 3 components, others with 4.

I am looking for a drop-in replacement for `torch.nn.ReLU` and `torch.nn.functional.relu()`. Can anyone provide that?
